src/sqlitestorage.cpp

```diff
@@ -368,202 +146,50 @@ bool SqliteStorage::load(const QString &uid)
-            query1 = SELECT_COMPONENTS_BY_DATE_START;
-            qsize1 = sizeof(SELECT_COMPONENTS_BY_DATE_START);
-            SL3_prepare_v2(d->mDatabase, query1, qsize1, &stmt1, NULL);
-            secsStart = d->mFormat->toOriginTime(loadStart);
-            SL3_bind_int64(stmt1, index, secsStart);
-            SL3_bind_int64(stmt1, index, secsStart);
-        } else if (loadEnd.isValid()) {
-            query1 = SELECT_COMPONENTS_BY_DATE_END;
-            qsize1 = sizeof(SELECT_COMPONENTS_BY_DATE_END);
-            SL3_prepare_v2(d->mDatabase, query1, qsize1, &stmt1, NULL);
-            secsEnd = d->mFormat->toOriginTime(loadEnd);
-            SL3_bind_int64(stmt1, index, secsEnd);
-        } else {
-            query1 = SELECT_COMPONENTS_ALL;
-            qsize1 = sizeof(SELECT_COMPONENTS_ALL);
-            SL3_prepare_v2(d->mDatabase, query1, qsize1, &stmt1, NULL);
-        }
-        count = d->loadIncidences(stmt1);
+        bool loadAllRecurringIncidences = !isRecurrenceLoaded();
+        QHash<QString, Incidence::List> hash;
+        success = d->mBackend.incidences(&hash, loadStart, loadEnd, loadAllRecurringIncidences) && d->addIncidences(hash);
 
-        if (count >= 0) {
+        if (success) {
             addLoadedRange(loadStart.date(), loadEnd.date());
         }
-        if (loadStart.isNull() && loadEnd.isNull()) {
-            setIsRecurrenceLoaded(count >= 0);
+        if (loadAllRecurringIncidences) {
+           setIsRecurrenceLoaded(success);
         }
-    } else {
-        count = 0;
     }
-error:
-    d->mIsLoading = false;
 
-    return count >= 0;
+    return success;
 }
 
 bool SqliteStorage::loadNotebookIncidences(const QString &notebookUid)
 {
-    if (!d->mDatabase) {
-        return false;
-    }
-
-    int rv = 0;
-    int count = -1;
-    d->mIsLoading = true;
-
-    const char *query1 = NULL;
-    int qsize1 = 0;
-
-    sqlite3_stmt *stmt1 = NULL;
-    int index = 1;
-    QByteArray u;
-
-    if (!notebookUid.isEmpty()) {
-        query1 = SELECT_COMPONENTS_BY_NOTEBOOKUID;
-        qsize1 = sizeof(SELECT_COMPONENTS_BY_NOTEBOOKUID);
-
-        SL3_prepare_v2(d->mDatabase, query1, qsize1, &stmt1, NULL);
-        u = notebookUid.toUtf8();
-        SL3_bind_text(stmt1, index, u.constData(), u.length(), SQLITE_STATIC);
-
-        count = d->loadIncidences(stmt1);
-    }
-error:
-    d->mIsLoading = false;
-
-    return count >= 0;
+    Incidence::List list;
+    return d->mBackend.incidences(&list, notebookUid) && d->addIncidences(list, notebookUid);
 }
 
-bool SqliteStorage::Private::loadRecurringIncidences()
+bool SqliteStorage::search(const QString &key, QStringList *identifiers, int limit)
 {
-    if (!mDatabase) {
-        return false;
-    }
-
-    if (mStorage->isRecurrenceLoaded()) {
+    QHash<QString, Incidence::List> hash;
+    QHash<QString, QStringList> ids;
+    if (d->mBackend.search(&hash, &ids, key, limit) && d->addIncidences(hash)) {
+        if (identifiers) {
```

Comment:
This method was earlier assuming identifiers non-null. Doing search without way to report findings would be silly? Could just assume non-null or then have a sanity check returning false at the beginning?